src/Servers/IIS/IIS/src/StartupHook.cs

```diff
@@ -0,0 +1,77 @@
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security.Claims;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting.Views;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.AspNetCore.Server.IIS;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.StackTrace.Sources;
+
+internal class StartupHook
+{
+    /// <summary>
+    /// Startup hooks are pieces of code that will run before a users program main executes
+    /// See: https://github.com/dotnet/core-setup/blob/master/Documentation/design-docs/host-startup-hook.md
+    /// </summary>
+    public static void Initialize()
+    {
+        if (Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT") != "Development")
+        {
+            return;
+        }
+
+        AppDomain.CurrentDomain.UnhandledException += (sender, eventArgs) =>
+        {
+            var exception = (Exception)eventArgs.ExceptionObject;
+
+            // Get the content root from IIS.
+            var iisConfigData = NativeMethods.HttpGetApplicationProperties();
+            var contentRoot = iisConfigData.pwzFullApplicationPath.TrimEnd(Path.DirectorySeparatorChar);
+
+            var model = new ErrorPageModel
+            {
+                RuntimeDisplayName = RuntimeInformation.FrameworkDescription
+            };
+
+            var systemRuntimeAssembly = typeof(System.ComponentModel.DefaultValueAttribute).GetTypeInfo().Assembly;
+            var assemblyVersion = new AssemblyName(systemRuntimeAssembly.FullName).Version.ToString();
+            var clrVersion = assemblyVersion;
+            model.RuntimeArchitecture = RuntimeInformation.ProcessArchitecture.ToString();
+            var currentAssembly = typeof(ErrorPage).GetTypeInfo().Assembly;
+            model.CurrentAssemblyVesion = currentAssembly
+                .GetCustomAttribute<AssemblyInformationalVersionAttribute>()
+                .InformationalVersion;
+            model.ClrVersion = clrVersion;
+            model.OperatingSystemDescription = RuntimeInformation.OSDescription;
+
+            var exceptionDetailProvider = new ExceptionDetailsProvider(
+                new PhysicalFileProvider(contentRoot),
+                sourceCodeLineCount: 6);
+
+            model.ErrorDetails = exceptionDetailProvider.GetDetails(exception);
+
+            var errorPage = new ErrorPage(model);
+
+            var stream = new MemoryStream();
+            // Sync over async here, but you can't have async code in startup hooks.
+            errorPage.ExecuteAsync(stream).GetAwaiter().GetResult();
+
+            // Get the raw content and set the error page.
+            stream.Position = 0;
+            var content = stream.ToArray();
+
+            NativeMethods.HttpSetStartupErrorPageContent(content);
```

Comment:
Is it possible to be running on a version of ANCM without this export? What happens in that case?

Reply:
Possible yes, but it would require doing crazy things. Both the managed and native dlls are in the same package, so you can't reference different versions if you tried. It would require you copy a different inprocess native dll to the published/build directory. In that case, this would throw. Trying to cover this case is pointless. 

The only thing I do want to protect against is making sure we are actually running inprocess in the startup hook.

Reply:
OK. So in practice, this should never throw